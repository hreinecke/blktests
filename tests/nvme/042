#!/bin/bash
# SPDX-License-Identifier: GPL-3.0+
# Copyright (C) 2021 Hannes Reinecke, SUSE Labs
#
# Test bi-directional authentication

. tests/nvme/rc

DESCRIPTION="Test bi-directional authentication"
QUICK=1

requires() {
	_nvme_requires
	_have_modules loop
	_require_nvme_trtype_is_fabrics
}


test() {
	local port
	local subsys="blktests-subsystem-1"
	local hostid="$(uuidgen)"
	local hostnqn="nqn.2014-08.org.nvmexpress:uuid:${hostid}"
	local scratch="/tmp/blktest-ns1.img"
	local hostkey
	local ctrlkey
	local ctrldev

	echo "Running ${TEST_NAME}"

	hostkey="$(nvme gen-dhchap-key -n ${subsys} 2> /dev/null)"
	if [ $? -ne 0 ] ; then
		echo "nvme gen-dhchap-key command missing"
		return 1
	fi

	ctrlkey="$(nvme gen-dhchap-key -n ${subsys} 2> /dev/null)"
	if [ $? -ne 0 ] ; then
		echo "nvme gen-dhchap-key command missing"
		return 1
	fi

	_setup_nvmet

	truncate -s 512M "${scratch}"

	port="$(_create_nvmet_port "${nvme_trtype}")"

	_create_nvmet_subsystem "${subsys}" "${scratch}"
	_add_nvmet_subsys_to_port "${port}" "${subsys}"
	_create_nvmet_host "${subsys}" "${hostnqn}" "${hostkey}" "${ctrlkey}"

	_set_nvmet_dhgroup "${hostnqn}" "ffdhe2048"

	# Step 1: Connect with host authentication only
	echo "Test host authentication"
	_nvme_connect_subsys "${nvme_trtype}" "${subsys}" \
			     "${def_traddr}" "${def_trsvcid}" \
			     "${hostnqn}" "${hostid}" \
			     "${hostkey}"

	ctrldev=$(_find_nvme_dev "${subsys}")
	if [ -z "$ctrldev" ] ; then
		echo "nvme controller not found"
	fi

	_nvme_disconnect_subsys "${subsys}"

	# Step 2: Connect with host authentication
	# and invalid ctrl authentication
	echo "Test host authentication and invalid ctrl authentication"
	_nvme_connect_subsys "${nvme_trtype}" "${subsys}" \
			     "${def_traddr}" "${def_trsvcid}" \
			     "${hostnqn}" "${hostid}" \
			     "${hostkey}" "${hostkey}"
	ctrldev=$(_find_nvme_dev "${subsys}1")
	if [ -n "${ctrldev}" ] ; then
		echo "nvme controller found!"
		_nvme_disconnect_subsys "${ctrldev}"
	fi

	# Step 3: Connect with host authentication
	# and valid ctrl authentication
	echo "Test host authentication and valid ctrl authentication"
	_nvme_connect_subsys "${nvme_trtype}" "${subsys}" \
			     "${def_traddr}" "${def_trsvcid}" \
			     "${hostnqn}" "${hostid}" \
			     "${hostkey}" "${ctrlkey}"

	ctrldev=$(_find_nvme_dev "${subsys}")
	if [ -z "$ctrldev" ] ; then
		echo "nvme controller not found"
	fi

	_nvme_disconnect_subsys "${subsys}"

	_remove_nvmet_subsystem_from_port "${port}" "${subsys}"
	_remove_nvmet_subsystem "${subsys}"

	_remove_nvmet_port "${port}"

	_remove_nvmet_host "${hostnqn}"

	rm ${scratch}

	echo "Test complete"
}
