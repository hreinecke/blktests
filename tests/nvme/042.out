Running nvme/042
Test host authentication
NQN:blktests-subsystem-1 disconnected 1 controller(s)
Test host authentication and invalid ctrl authentication
no controller found
Test host authentication and valid ctrl authentication
NQN:blktests-subsystem-1 disconnected 1 controller(s)
Test complete
